nntrainer/tensor/tensor_pool.cpp

```diff
@@ -0,0 +1,214 @@
+  const TensorDim dim, const std::vector<unsigned int> &exec_order,
+
+  TensorLifespan lifespan, const std::string &name) {
+  if (pool.find(name) == pool.end())
+    throw std::invalid_argument("Requested tensor not found");
+
+  auto &spec = pool[name];
+  if (spec.tensor->getDim() != dim)
+    throw std::invalid_argument("Request tensor dimension mismatch");
+
+  spec.exec_order.insert(spec.exec_order.end(), exec_order.begin(),
+                         exec_order.end());
+  spec.lifespan = enum_class_or<TensorLifespan>(spec.lifespan, lifespan);
+
+  return pool[name].tensor.get();
+}
+
+/**
+ * @brief finalize the requested tensors
+ *
+ * @details finalize the requested tensors, request memory for them and plan
+ * layout for their allocations.
+ */
+void TensorPool::finalize(const MemoryPlanner &planner,
+                          unsigned int start_order, unsigned int end_order) {
+  unsigned int bytes_requested = 0;
+  for (auto &entry : pool) {
+    auto &spec = entry.second;
+
+    if (spec.exec_order.empty())
+      continue;
+
+    /** 1. create the validity ranges for the all the requested tensors */
+    unsigned int validity_start =
+      *std::min_element(spec.exec_order.begin(), spec.exec_order.end());
+    unsigned int validity_end =
+      *std::max_element(spec.exec_order.begin(), spec.exec_order.end());
+
+    /**
+     * use lifespan to update the validity.
+     * if the validity is long term, the tensor must stay valid for the
+     * complete duration.
+     */
+    if (isTensorLongTerm(spec.lifespan)) {
+      validity_start = start_order;
+      validity_end = end_order;
+    }
+
+    /** 2. for each tensor request if it is in the provided range */
+    if (validity_end < start_order || validity_start > end_order)
+      continue;
+    validity_start = std::max(validity_start, start_order);
+    validity_end = std::max(validity_end, end_order);
+
+    /**
+     * 3. requestMemory for all the tensors and set their tokens
+     * @note +1 is to make the validity_end exlusive in the interval range
+     */
+    spec.token = mem_pool.requestMemory(spec.tensor->bytes(), validity_start,
+                                        validity_end + 1);
+    bytes_requested += spec.tensor->bytes();
+  }
+
+  /** 4. finalizeLayout for the memory pool. */
+  if (bytes_requested > 0)
+    mem_pool.planLayout(planner);
+}
+
+/**
+ * @brief Set the batch size for the inputs/outputs of the layers
+ */
+void TensorPool::setBatchSize(const std::string &name, unsigned int batch) {
+  if (pool.find(name) == pool.end())
+    throw std::invalid_argument("Requested tensor not found");
+
+  pool[name].tensor->updateBatch(batch);
+}
+
+/**
+ * @brief Allocate memory for all the managed tensors
+ */
+void TensorPool::allocate() {
+  mem_pool.allocate();
+
+  /** set the pointers using the token for all the tensors */
+  for (auto &entry : pool) {
+    auto &spec = entry.second;
+    spec.tensor->setData(mem_pool.getMemory(spec.token));
```

Comment:
wonder if source tensor is going away?

Reply:
can you elaborate? If you are referring to src_tensor in tensor.h, then eventually that would be replaced with tensor_pool interface. But it would require all tensors to be created via tensor_pool, which would take some time.

Reply:
I was wondering about the `src_tensor` in tensor.h will eventually go away. Thanks :)

Reply:
Yes, I think it can eventually go away, and will be good to make tensor objects lighter.

Reply:
I think it isn't easy to remove src_tensor. Because some of the ops ( as in RNN or GRU Layer ) is based on src_tensor. We need to check if it is ok when we remove src_tensor.

Reply:
Yeah, we will have to check with all the existing usages when working on it.